src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs

```diff
@@ -343,7 +357,141 @@ public void Stop()
+                    _spanIdSet = true;
+                }
+                return ref _spanId;
+            }
+        }
+
+        /// <summary>
+        /// If the Activity has the W3C format, this returns the ID for the TraceId part of the Id.  
+        /// Otherwise it returns a zero TraceId. 
+        /// </summary>
+        public ref TraceId TraceId
+        {
+            get
+            {
+                if (!_traceIdSet)
+                {
+                    if (IdFormat == ActivityIdFormat.W3C)
+                        _traceId = new TraceId(RootId);
+                    _traceIdSet = true;
+                }
+                return ref _traceId;
+            }
+        }
+
+        /// <summary>
+        /// Returns the format for the ID.   
+        /// </summary>
+        public ActivityIdFormat IdFormat { get; private set; }
+
+        /* static state (configuration) */
+        /// <summary>
+        /// Activity tries to use the same format for IDs as its parent.
+        /// However if the activity has no parent, it has to do something.   
+        /// This determines the default format we use.  
+        /// </summary>
+        public static ActivityIdFormat DefaultIdFormat
+        {
+            get { return s_DefaultIdFormat; }
+            set
+            {
+                if (!(ActivityIdFormat.Hierarchical <= value && value <= ActivityIdFormat.W3C))
+                    throw new ArgumentException($"value must be a valid ActivityIDFormat value");
+                s_DefaultIdFormat = value;
+
+            }
+        }
+
+        /// <summary>
+        /// Normally if the ParentID is defined, the format of that is used to determine the
+        /// format used by the Activity.   However if ForceDefaultFormat is set to true, the
+        /// ID format will always be the DefaultIdFormat even if the ParentID is define and is
+        /// a different format. 
+        /// </summary>
+        public bool ForceDefaultIdFormat { get; set; }
+
         #region private 
+        /// <summary>
+        /// Returns true if 'id' has the format of a WC3 id see https://w3c.github.io/trace-context
+        /// </summary>
+        private static bool IsW3CId(string id)
+        {
+            // A W3CId is  
+            //  * 2 chars Version
+            //  * 32 chars traceId
+            //  * 16 chars spanId
+            //  * 2 chars flags 
+            //  * 3 chars - separators.  
+            //  = 55 chars (see https://w3c.github.io/trace-context)
+            // We require that all non-WC3IDs NOT start with a digit.  
+            return id.Length == 55 && char.IsDigit(id[0]);
+        }
+        private static bool IsW3CId(Span<byte> id)
+        {
+            return id.Length == 55 && char.IsDigit((char) id[0]);
+        }
+
+        private string GenerateW3CId()
+        {
+            string newSpanId = Guid.NewGuid().ToString("n").Substring(16, 16);
+            if (ParentId != null)
```

Comment:
I believe if ParentId is hierarchical, but W3C is forced, this code does not produce valid W3C traceparent.
This scenario is expected when AspNetCore 2.1 will be used with new DiagnosticSource.